File: Scripts/RMSD/RMSD_analysis.R
cat("Running: RMSD_analysis.R\n")

############################################################################################

# Installing and Loading Packages
required_packages <- c("dplyr", "stringr", "tidyr","readr", "ggplot2", "dendextend", "RColorBrewer", "tibble")

not_installed <- required_packages[!(required_packages %in% installed.packages()[ , "Package"])]

# Install missing packages
if(length(not_installed)) install.packages(not_installed)

# Load all required packages
invisible(lapply(required_packages, require, character.only = TRUE))
  
# Cleaning Environment
rm(list = ls())

############################################################################################

### Reading in configuration from run_analysis.sh ###

# Function to read configuration from stdin
read_config <- function() {
  config <- list()
  lines <- readLines(file("stdin"))
  for (line in lines) {
    parts <- strsplit(line, "=")[[1]]
    config[[parts[1]]] <- parts[2]
  }
  return(config)
}

# Read the configuration from stdin
config <- read_config()

# Convert 'custom_cut_height' parameter to numeric
custom_cut_height <- as.numeric(config$custom_cut_height)

# Check the custom_cut_height
print(paste("Custom cut height:", custom_cut_height))

############################################################################################

output_dir <- file.path("Output", "RMSD")

data_path <- file.path("Output", "RMSD", "data")

single_pdb_dir <- file.path("Output", "RMSD", "single_reference")

# Creating Directories
if(!dir.exists(single_pdb_dir)) dir.create(single_pdb_dir)

############################################################################################

# Read in df
df <- read.csv(file.path(data_path, "pairwise_rmsd.csv"))

############################################################################################

### Removing Low Resolution Residues ###

# Importing data
high_resolution_residues <- read.csv("Output/Validation/high_resolution_residues.csv")

high_res_pdb_ids <- unique(high_resolution_residues$pdb_id)

filtered_df <- df %>% filter(ref_name %in% high_res_pdb_ids) %>%
  filter((mob_name %in% high_res_pdb_ids))

############################################################################################

# Getting amyloid names from first chain only
amyloid_names <- unique(filtered_df$ref_name)

for (i in amyloid_names) {
    
  x <- filtered_df %>% filter(ref_name == i)
  
  # Plotting RMSD for each reference
  ggplot(x, aes(x = reorder(mob_name, rmsd), y = rmsd)) +
    geom_point(size = 4) +
    ggtitle(paste(i, " - Chain Comparison", sep = "")) +
    ylab("RMSD") +
    xlab("PDB") +
    scale_y_continuous(expand = c(0, 0), limits = c(0, max(filtered_df$rmsd) * 1.05),
                       breaks = seq(0,1000,5)) +
    theme_bw() +
    theme(panel.grid.major.y = element_line(colour = "grey30", linewidth = 0.25,
                                            linetype = "dashed"),
          panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
          plot.title = element_text(size = 20, face = "bold", colour = "black", hjust = 0.5),
          axis.title = element_text(size = 18, face = "bold", colour = "black"),
          axis.text.x = element_text(size = 13, colour = "black",
                                     angle = 90, vjust = 0.4, hjust = 1),
          axis.text.y = element_text(size = 14, colour = "black")) 
    
  ggsave(plot = last_plot(), file = file.path(single_pdb_dir, paste0(i, "_RMSD.png")), 
         width = 20, height = 8)
  
}

############################################################################################

# Wide format data
df_wide <- pivot_wider(
  filtered_df,
  names_from = mob_name,
  values_from = rmsd
)

names(df_wide)[names(df_wide) == "ref_name"] <- "pdb_id"

# Removing reference columns
cluster_data <- subset(df_wide, select = setdiff(names(df_wide), c("pdb_id")))

# Setting NA values to the mean + 3SD - This occurs when two PDBs have no overlapping residues 
NA_distance <- mean(as.matrix(cluster_data), na.rm = TRUE) + 3*sd(as.matrix(cluster_data), na.rm = TRUE)

cluster_data[is.na(cluster_data)] <- NA_distance

# Scaling Cluster Data (makes mean of all columns = 0 and SD = 0-1)
scaled_cluster_data <- as.data.frame(scale(cluster_data))

# Saving cluster data
write.csv(scaled_cluster_data, file = file.path(data_path, "scaled_cluster_data.csv"), row.names = FALSE)

############################################################################################

################################
### Clustering Based On RMSD ###
################################

# Calculating Distance
cluster_dist <- dist(scaled_cluster_data, method = "euclidean")

# Clustering
hc_average <- hclust(cluster_dist, method = "average")


### Scree Plot to assigning cut height ###
scree_data <- hc_average$height %>%
              as_tibble() %>%
              add_column(groups = length(hc_average$height):1) %>%
              rename(height=value)

# Saving Scree data
write.csv(scree_data, file = file.path(data_path, "scree_data.csv"), row.names = FALSE)

# Plotting 
ggplot(scree_data, aes(x=groups, y=height)) +
       geom_point(size = 1.5) +
       geom_line(linewidth = 0.5) +
       annotate(geom = "segment", linewidth = 0.5,, colour = "red",
                x = -Inf, xend = Inf,
                y = custom_cut_height, yend = custom_cut_height) +
  scale_x_continuous(expand = c(0,0.75), breaks = seq(0,9999,10)) +
  scale_y_continuous(#limits = c(0,30),
                     expand = c(0,0.75)) +
  labs(x = "Groups", y = "Euclidean Distance") +
  theme_bw() +
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
        panel.grid.major = element_line(colour = "grey", linewidth = 0.5, linetype = "dashed"),
        panel.grid.minor = element_line(colour = "grey", linewidth = 0.5, linetype = "dashed"),
        axis.title = element_text(size = 16, colour = "black", face = "bold"),
        axis.text = element_text(size = 14, colour = "black"))

# Saving the scree plot
ggsave(plot = last_plot(), filename = file.path(output_dir, "Scree_plot.png"), height = 4, width = 8)



# Making labels
pdb_id_labels <- df_wide$pdb_id[hc_average$order]

# Creating a dataframe to store the order of PDBs in the dendrogram
dendrogram_order <- data.frame(pdb_id = pdb_id_labels,
                      order = seq(1,length(pdb_id_labels),1))

# Saving Dendrogram Order
write.csv(dendrogram_order, file = file.path(data_path, "dendrogram_order.csv"), row.names = FALSE)


if (is.na(custom_cut_height) || custom_cut_height <= 0 || is.null(custom_cut_height)) {
  
  # Setting cut_height to the mean distance
  cut_height <- mean(cluster_dist)
  print(paste0("Using mean Euclidean distance as the cut height: ", cut_height))
  
} else {
  
  cut_height = custom_cut_height
  print(paste0("Using custom cut height: ", cut_height))

}

# Extracting pdb_id from each cluster group
clusters <- cutree(hc_average, h = cut_height)
grouped_pdbs <- split(df_wide$pdb_id, clusters)

# Convert the list to a dataframe
cluster_groups <- do.call(rbind, lapply(names(grouped_pdbs), function(name) {
  data.frame(group = name, pdb_id = grouped_pdbs[[name]], stringsAsFactors = FALSE)
}))

# Ordering the colour column by the RMSD Dendrogram order
pdb_id_labels_factor <- factor(pdb_id_labels, levels = pdb_id_labels)
cluster_groups <- cluster_groups[order(factor(cluster_groups$pdb_id, levels = levels(pdb_id_labels_factor))), ]

# Ordering cluster_groups by dendrogram order 
cluster_groups <- cluster_groups[match(pdb_id_labels, cluster_groups$pdb_id), ]

tmp <- as.character(unique(cluster_groups$group))

cluster_groups_sorted <- cluster_groups

for (i in 1:length(tmp)) {
  
  cluster_groups_sorted$group[cluster_groups_sorted$group == tmp[i]] <- paste0(i,".")  
  
}

# Removing temporary "."
cluster_groups_sorted$group <- gsub("\\.", "", cluster_groups_sorted$group)


# Saving Cluster Groups
write.csv(cluster_groups_sorted, file = file.path(data_path, "RMSD_cluster_groups.csv"), row.names = FALSE)

# Making group column numeric
cluster_groups$group <- as.numeric(cluster_groups$group)

# Setting branch colours 
colours <- c("red", "blue", "gold2", "purple", "darkorange2", "cyan", "green", "pink", "chocolate4", 
             "darkgreen", "magenta", "deepskyblue", "darkviolet", "coral", "darkslategray")

# Extend the color vector to match groups
branch_colours <- rep(colours, length.out = as.numeric(max(cluster_groups$group)))

# Using group number to index the colour palette
cluster_groups$colour <- branch_colours[as.numeric(cluster_groups$group)]

# Taking only the unique colour values as the plot requires one value for each group (not for each line)
group_colours <- unique(cluster_groups$colour)


# Plotting
avg_dend_obj <- as.dendrogram(hc_average)
avg_col_dend <- color_branches(avg_dend_obj, h = cut_height, col = group_colours)
labels(avg_col_dend) <- pdb_id_labels
avg_col_dend <- set(avg_col_dend, "branches_lwd", 2.5) %>% set("labels_cex", 1)

# Save the cluster plot as a PNG file
png(file.path(output_dir, "RMSD_cluster_dendrogram.png"), width = 6000, height = 1500, res = 300)
plot(avg_col_dend, ylab = "Euclidean Distance", xlab = "PDB ID",
     cex.lab = 1.2, font.lab = 2, cex.axis = 1)
abline(h = cut_height, col = "grey30", lty = 2, lw = 2)
title("Cluster Dendrogram - RMSD")
dev.off()

#############################
### RMSD Heatmap Plotting ###
#############################

# RMSD = Root Mean Squared Distance

# Converting dataframe into long format for plotting
mean_distance_heatmap <- df_wide %>% pivot_longer(!pdb_id, names_to = "comparison", values_to = "mean_distance")

# Ordering by cluster order
mean_distance_heatmap$pdb_id <- factor(mean_distance_heatmap$pdb_id, levels = pdb_id_labels)
mean_distance_heatmap$comparison <- factor(mean_distance_heatmap$comparison, levels = pdb_id_labels)

# Heatmap plot
ggplot(mean_distance_heatmap, aes(x = comparison, y = pdb_id, fill= mean_distance)) +
  geom_tile() +
  ggtitle("RMSD Heatmap") +
  ylab("PDB") +
  xlab("PDB") +
  #scale_fill_gradient(low="black", high="red", na.value = "white") +
  scale_fill_gradientn(colors = c("red", "black", "blue")) +
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
        plot.title = element_text(size = 20, face = "bold", colour = "black", hjust = 0.5),
        axis.text.y = element_text(size = 12, colour = "black", face = "bold"),
        axis.text.x = element_text(size = 12, colour = "black", face = "bold", angle = 90,
                                   vjust = 0.4, hjust = 1),
        axis.title=element_text(size = 25, face = "bold"),
        axis.line.x = element_line(color="black", linewidth = 1.5),
        axis.line.y = element_line(color="black", linewidth = 1.5),
        legend.title = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 16, face = "bold")) +
  labs(fill = "RMSD")

# Saving the heatmap
ggsave(plot = last_plot(), file = file.path(output_dir, "RMSD_heatmap.png"),
      width = 18, height = 15)

# Saving data
write.csv(mean_distance_heatmap, file = file.path(data_path, "RMSD_heatmap.csv"), row.names = FALSE)


# Cleaning Environment
#rm(list = ls())
invisible(gc())

File: Scripts/RMSD/RMSD_analysis.py
import os
import sys
import pandas as pd
import numpy as np
from scipy.spatial.distance import pdist
from scipy.cluster.hierarchy import linkage, dendrogram, leaves_list, fcluster
import matplotlib.pyplot as plt
from plotnine import *

# Setting up paths
output_dir = os.path.join("Output", "RMSD")
data_path = os.path.join(output_dir, "data")
single_pdb_dir = os.path.join(output_dir, "single_reference")

# Create output directories if they don't exist
if not os.path.exists(single_pdb_dir):
    os.mkdir(single_pdb_dir)

### Reading in command line arguments ###

if len(sys.argv) > 1: 
    try: 
        custom_cut_height = int(sys.argv[1]) 
        print(f"Cut Height: {custom_cut_height}") 
    except ValueError: 
        print("Warning: invalid cut height provided — defaulting to mean Euclidean distance") 
        custom_cut_height = 0 
else: 
    custom_cut_height = 0 
    print("No cut height provided, defaulting to mean Euclidean distance")


# Importing data
df = pd.read_csv(os.path.join(data_path, "pairwise_rmsd.csv"), sep=",")
high_resolution_residues = pd.read_csv(os.path.join("Output", "Validation", "high_resolution_residues.csv"), sep=",")


### Removing low resolution PDBs from df ###

amyloid_names = high_resolution_residues['pdb_id'].astype(str).unique()

filtered_df = df[
    df['ref_name'].isin(amyloid_names) &
    df['mob_name'].isin(amyloid_names)
].copy()

##############################################################################

### Plotting the RMSD for each reference structure ###

for i in amyloid_names:
    
    # Selecting data for the current reference structure
    x = filtered_df[filtered_df['ref_name'] == i]

    # Sorting RMSD in ascending order
    x = x.sort_values(by='rmsd', ascending=True)

    # Making pdb_id an ordered categorical variable for plotting
    x["mob_name"] = pd.Categorical(
        x["mob_name"],
        categories=x["mob_name"].tolist(),
        ordered=True
    )

    p = (
        ggplot(x, aes(x='mob_name', y='rmsd')) 
        + geom_point(size = 4)
        + labs(title=f'RMSD values for reference structure {i}', 
               x='PDB ID', 
               y='RMSD (Å)')
        + scale_y_continuous(limits=(0, np.max(x['rmsd']) * 1.05), expand = (0, 0, 0.1, 0), 
                             breaks=range(0, 999999, 5)) 
        + theme_bw()
        + theme(
            panel_grid_major_y = element_line(color = "grey", size = 0.25, linetype = "dashed"),
            panel_grid_major_x = element_blank(),
            panel_border = element_rect(color = "black", size = 1.5),
            plot_title = element_text(size = 20, face = "bold", ha = "center"),
            axis_title = element_text(size = 18, face = "bold"),
            axis_text_y = element_text(size = 14, colour = "black"),
            axis_text_x = element_text(size = 13, angle = 90, vjust = 0.5, hjust=1, colour = "black")
        )
    )

    # Saving plot
    p.save(os.path.join(single_pdb_dir, f'{i}_RMSD.png'), width=20, height=8, dpi=300)

##############################################################################

# Pivot the DataFrame to wide format
df_wide = filtered_df.pivot(index='ref_name', columns='mob_name', values='rmsd').reset_index()

# Rename 'ref_name' to 'pdb_id'
df_wide = df_wide.rename(columns={'ref_name': 'pdb_id'})

# Set 'pdb_id' as the index
cluster_data = df_wide.set_index('pdb_id')

### Handling PDBs with no overlapping residues by setting the RMSD to mean + 3*SD ###

# Compute mean + 3*SD for all numeric entries, ignoring NaN
NA_distance = np.nanmean(cluster_data.values) + 3 * np.nanstd(cluster_data.values)
# Replace NaNs with NA_distance
cluster_data = cluster_data.fillna(NA_distance)

# Scale the data (mean = 0, std = 1)
scaled_cluster_data = (cluster_data - cluster_data.mean()) / cluster_data.std()

# Save the scaled data to a CSV file
scaled_cluster_data.to_csv(os.path.join(data_path, "scaled_cluster_data.csv"))

############################################################################################

################################
### Clustering Based On RMSD ###
################################

# Calculate Euclidean distance matrix
cluster_dist = pdist(scaled_cluster_data, metric='euclidean')

# Perform hierarchical clustering using average linkage
hc_average = linkage(cluster_dist, method='average')


### Comparing number of clusters vs cut height ###

# Extract linkage heights
heights = hc_average[:, 2]  # third column of linkage matrix

# Create scree_data equivalent
scree_data = pd.DataFrame({
    "height": heights,
    "groups": np.arange(len(heights), 0, -1)
})

# Save to CSV
scree_data.to_csv(os.path.join(data_path, "scree_data.csv"), index=False)

# Plotting the scree plot
p = (
    ggplot(scree_data, aes(x='groups', y='height'))
    + geom_point(size=1.5) 
    + geom_line(size=0.5)
    + annotate("segment", x=1, xend=len(heights), y=float(custom_cut_height), yend=float(custom_cut_height), color="red", linetype="dashed", size=0.75)
    + scale_x_continuous(breaks=range(0, len(heights)+1, 10), expand = (0, 0.05))
    + scale_y_continuous(expand = (0, 0, 0.05, 0))
    + labs(title="", 
           x="Groups", 
           y="Euclidean Distance (Å)")
    + theme_bw()
    + theme(
        panel_border = element_rect(color = "black", size = 1.5),
        panel_grid_major = element_line(colour = "grey", size = 0.5, linetype = "dashed"),
        panel_grid_minor = element_line(colour = "grey", size = 0.5, linetype = "dashed"),
        axis_title = element_text(size = 16, face = "bold"),
        axis_text = element_text(size = 14, colour = "black")
    )
)

# Saving plot
p.save(os.path.join(output_dir, "scree_plot.png"), height=4, width=8, dpi=300)

##################
### Dendrogram ###
##################

# Get the dendrogram order
leaf_order = leaves_list(hc_average)

# Getting a list of PDB IDs in the order of the dendrogram leaves
pdb_id_labels = df_wide["pdb_id"].iloc[leaf_order].reset_index(drop=True)

# Create the dendrogram order DataFrame
dendrogram_order = pd.DataFrame({
    "pdb_id": pdb_id_labels,
    "order": np.arange(1, len(pdb_id_labels) + 1)
})

#### Cutting the dendrogram ###

# Setting the cut height
if float(custom_cut_height) == 0:
    cut_height = np.mean(cluster_dist)
    print(f"Using mean Euclidean distance as the cut height: {cut_height}")
else:
    cut_height = float(custom_cut_height)
    print(f"Using custom cut height: {cut_height}")

# Assign clusters based on the cut height
cluster_assignments = fcluster(hc_average, t=cut_height, criterion='distance')

# Create a DataFrame for cluster assignments
cluster_assignments = pd.DataFrame({
    "pdb_id": df_wide["pdb_id"],
    "cluster": cluster_assignments
})

# Merge dendrogram order with cluster assignments
cluster_groups = pd.merge(dendrogram_order, cluster_assignments, on="pdb_id")

# Identify the order in which clusters appear from left to right
unique_cluster_order = cluster_groups["cluster"].drop_duplicates().tolist()

# Create a mapping: first seen cluster -> 1, next -> 2, etc.
cluster_mapping = {old: new for new, old in enumerate(unique_cluster_order, start=1)}

# Apply the renumbering
cluster_groups["group"] = cluster_groups["cluster"].map(cluster_mapping)

# Remove the old cluster column
cluster_groups = cluster_groups.drop(columns=["cluster"])

# Save the final cluster groups to a CSV file
cluster_groups.to_csv(os.path.join(data_path, "RMSD_cluster_groups.csv"), index=False)

### Plotting the dendrogram ###

# plot dendrogram
plt.figure(figsize=(12,6))
d = dendrogram(hc_average, 
               labels=df_wide["pdb_id"].tolist(), 
               leaf_rotation=90, 
               color_threshold=cut_height,
               above_threshold_color="black")
plt.axhline(y=cut_height, color='grey', linestyle='--', linewidth=1.5)
plt.title("Hierarchical Clustering Dendrogram")
plt.xlabel("PDB ID")
plt.ylabel("Distance")
plt.tight_layout()
plt.savefig(os.path.join(output_dir, "RMSD_cluster_dendrogram.png"), dpi=300, bbox_inches="tight")

##############################################################################

#############################
### RMSD Heatmap Plotting ###
#############################

# Convert df_wide to long format
mean_distance_heatmap = df_wide.melt(
    id_vars='pdb_id',
    var_name='comparison',
    value_name='mean_distance'
)

# Specifying plot order based on RMSD cluster order stored in pdb_id_labels
mean_distance_heatmap['pdb_id'] = pd.Categorical(mean_distance_heatmap['pdb_id'], categories=pdb_id_labels, ordered=True)
mean_distance_heatmap['comparison'] = pd.Categorical(mean_distance_heatmap['comparison'], categories=pdb_id_labels, ordered=True)

# Applying the order to the DataFrame (so it is saved in the correct order in the CSV file)
mean_distance_heatmap = mean_distance_heatmap.sort_values(['pdb_id', 'comparison']).reset_index(drop=True)

# Saving the mean distance heatmap data to a CSV file
mean_distance_heatmap.to_csv(os.path.join(data_path, "RMSD_heatmap.csv"), index=False)

# Heatmap plot
p = (
    ggplot(mean_distance_heatmap, aes(x='comparison', y='pdb_id', fill='mean_distance'))
    + geom_tile(color='white')
    + scale_fill_gradient(low='white', high='red', name='Mean RMSD (Å)', limits=(0, np.max(mean_distance_heatmap['mean_distance']) * 1.05))
    + labs(title='Mean Pairwise RMSD Heatmap', x='PDB ID', y='PDB ID')
    + theme_minimal()
    + theme(
        panel_border = element_rect(color="black", size=1.5),
        panel_grid_major = element_blank(),
        panel_grid_minor = element_blank(),
        plot_title = element_text(size=20, face="bold", ha="center"),
        axis_title = element_text(size=25, face="bold"),
        axis_text_x = element_text(size=12, colour="black", angle=90, vjust=0.5, hjust=1),
        axis_text_y = element_text(size=12, colour="black"),
        legend_title = element_text(size=20, face="bold"),
        legend_text = element_text(size=16)
    )
)

# Saving plot
p.save(os.path.join(output_dir, "RMSD_heatmap.png"), width=18, height=15, dpi=300)

